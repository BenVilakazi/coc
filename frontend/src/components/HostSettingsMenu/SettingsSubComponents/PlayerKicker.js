import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import OptionList from './OptionList';
import { HostContext } from '../../../contexts/HostContext/HostContext';

const propTypes = {
  accordionState: PropTypes.string.isRequired,
  onClickActions: PropTypes.shape({
    open: PropTypes.func.isRequired,
    enabled: PropTypes.func.isRequired,
    disabled: PropTypes.func.isRequired,
  }).isRequired,
};

function PlayerKicker({ accordionState, onClickActions }) {
  const {
    state: { playerIDs, players },
    dispatch,
  } = useContext(HostContext);

  // this variable is needed for the OptionList API; it needs the player names
  // in an array of strings
  const playerList = playerIDs
    .map((playerId) => ({
      text: players[playerId].name.toUpperCase(),
      value: playerId,
    }))
    .sort((a, b) => b > a);

  function kickPlayer(targetId) {
    const czarId =
      playerIDs.length > 0 &&
      playerIDs.find((playerId) => players[playerId].isCzar);

    dispatch({
      type: 'KICK_PLAYER',
      payload: { playerId: targetId, czarId },
    });
  }

  return (
    <OptionList
      listContent={playerList}
      state={accordionState}
      // if you get a "cannot read property 'undefined' of undefined" error on
      // the below line and a bunch of propType errors in your console check
      // that you assigned it the right component type
      onClick={onClickActions[accordionState]}
      onItemClick={kickPlayer}
      openText="KICK WHO?"
      closedText="KICK PLAYER"
    />
  );
}

PlayerKicker.propTypes = propTypes;

export default PlayerKicker;
