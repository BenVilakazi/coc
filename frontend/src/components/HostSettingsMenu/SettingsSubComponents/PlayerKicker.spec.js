import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import PlayerKicker from './PlayerKicker';
import HostProvider, {
  HostContext,
} from '../../../contexts/HostContext/HostContext';

jest.mock(
  './OptionList', // eslint-disable-next-line react/prop-types
  () => ({ listContent, onClick, onItemClick }) => {
    // eslint-disable-next-line react/prop-types
    function MakeListItemButton({ text, value }) {
      return (
        <button
          type="button"
          key={text + value}
          onClick={() => onItemClick(value)}
        >
          {text}
        </button>
      );
    }

    return (
      <div>
        <button type="button" onClick={onClick}>
          MAIN BUTTON
        </button>
        {
          // eslint-disable-next-line react/prop-types
          listContent.map((item) => MakeListItemButton(item))
        }
      </div>
    );
  },
);

describe('PlayerKicker', () => {
  describe('rendering', () => {
    it('matches the snapshot', () => {
      const tree = renderer
        .create(
          <HostProvider>
            <PlayerKicker
              accordionState="enabled"
              onClickActions={{
                open: () => {},
                enabled: () => {},
                disabled: () => {},
              }}
            />
          </HostProvider>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('upcases player names before passing them to be rendered', () => {
      const state = {
        players: {
          player1: {
            name: 'foo',
            submittedCards: [],
          },
        },
        playerIDs: ['player1'],
      };
      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <PlayerKicker
            accordionState="open"
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostContext.Provider>,
      );

      expect(screen.queryByText('foo')).not.toBeInTheDocument();
      expect(screen.getByText('FOO')).toBeInTheDocument();
    });
  });

  describe('functionality', () => {
    it('calls the onClickActions open callback when the main button is click and state is open', () => {
      const mockOpenOnClick = jest.fn();
      const mockEnabledOnClick = jest.fn();
      const mockDisabledOnClick = jest.fn();

      render(
        <HostProvider>
          <PlayerKicker
            accordionState="open"
            onClickActions={{
              open: mockOpenOnClick,
              enabled: mockEnabledOnClick,
              disabled: mockDisabledOnClick,
            }}
          />
        </HostProvider>,
      );

      userEvent.click(screen.getByRole('button', { name: 'MAIN BUTTON' }));

      expect(mockOpenOnClick).toHaveBeenCalled();
      expect(mockEnabledOnClick).not.toHaveBeenCalled();
      expect(mockDisabledOnClick).not.toHaveBeenCalled();
    });

    it('calls the onClickActions enabled callback when the main button is click and state is enabled', () => {
      const mockOpenOnClick = jest.fn();
      const mockEnabledOnClick = jest.fn();
      const mockDisabledOnClick = jest.fn();

      render(
        <HostProvider>
          <PlayerKicker
            accordionState="enabled"
            onClickActions={{
              open: mockOpenOnClick,
              enabled: mockEnabledOnClick,
              disabled: mockDisabledOnClick,
            }}
          />
        </HostProvider>,
      );

      userEvent.click(screen.getByRole('button', { name: 'MAIN BUTTON' }));

      expect(mockOpenOnClick).not.toHaveBeenCalled();
      expect(mockEnabledOnClick).toHaveBeenCalled();
      expect(mockDisabledOnClick).not.toHaveBeenCalled();
    });

    it('calls the onClickActions disabled callback when the main button is click and state is disabled', () => {
      const mockOpenOnClick = jest.fn();
      const mockEnabledOnClick = jest.fn();
      const mockDisabledOnClick = jest.fn();

      render(
        <HostProvider>
          <PlayerKicker
            accordionState="disabled"
            onClickActions={{
              open: mockOpenOnClick,
              enabled: mockEnabledOnClick,
              disabled: mockDisabledOnClick,
            }}
          />
        </HostProvider>,
      );

      userEvent.click(screen.getByRole('button', { name: 'MAIN BUTTON' }));

      expect(mockOpenOnClick).not.toHaveBeenCalled();
      expect(mockEnabledOnClick).not.toHaveBeenCalled();
      expect(mockDisabledOnClick).toHaveBeenCalled();
    });

    it('does not cause names to be rendered when no players are provided from the context', () => {
      const state = {
        players: {},
        playerIDs: [],
      };
      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <PlayerKicker
            accordionState="open"
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostContext.Provider>,
      );

      expect(screen.queryByText('FOO')).not.toBeInTheDocument();
      expect(screen.queryByText('BAR')).not.toBeInTheDocument();
    });

    it('gets players from the host context and passes their names to be rendered', () => {
      const state = {
        players: {
          player1: {
            name: 'FOO',
            submittedCards: [],
          },
          player2: {
            name: 'BAR',
            submittedCards: [],
          },
        },
        playerIDs: ['player1', 'player2'],
      };
      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <PlayerKicker
            accordionState="open"
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostContext.Provider>,
      );

      expect(screen.getByText('FOO')).toBeInTheDocument();
      expect(screen.getByText('BAR')).toBeInTheDocument();
    });

    it('does not dispatch when no name has been clicked', () => {
      const state = {
        players: {
          player1: {
            name: 'FOO',
            submittedCards: [],
          },
          player2: {
            name: 'BAR',
            submittedCards: [],
          },
        },
        playerIDs: ['player1', 'player2'],
      };
      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <PlayerKicker
            accordionState="open"
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostContext.Provider>,
      );

      expect(dispatch).not.toHaveBeenCalled();
    });

    it('dispatches with the correct type and payload when a name is clicked', () => {
      const state = {
        players: {
          player1: {
            name: 'FOO',
            submittedCards: [],
          },
          player2: {
            name: 'BAR',
            submittedCards: [],
          },
        },
        playerIDs: ['player1', 'player2'],
      };
      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <PlayerKicker
            accordionState="open"
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostContext.Provider>,
      );

      userEvent.click(screen.getByRole('button', { name: 'FOO' }));

      expect(dispatch).toHaveBeenCalledWith({
        type: 'KICK_PLAYER',
        payload: { playerId: 'player1' },
      });
    });
  });

  describe('propTypes', () => {
    it('logs an error when attempting to render without the accordionState Prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <PlayerKicker
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostProvider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the accordionState Prop as a non string', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <PlayerKicker
            accordionState
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostProvider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the onClickActions Prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <HostProvider>
            <PlayerKicker accordionState="open" />
          </HostProvider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the onClickActions Prop without the open key', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <PlayerKicker
            accordionState
            onClickActions={{
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostProvider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the onClickActions Prop without the enabled key', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <PlayerKicker
            accordionState
            onClickActions={{
              open: () => {},
              disabled: () => {},
            }}
          />
        </HostProvider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the onClickActions Prop without the disabled key', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <PlayerKicker
            accordionState
            onClickActions={{
              open: () => {},
              enabled: () => {},
            }}
          />
        </HostProvider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('does not log an error when all required props are given correctly', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <PlayerKicker
            accordionState="enabled"
            onClickActions={{
              open: () => {},
              enabled: () => {},
              disabled: () => {},
            }}
          />
        </HostProvider>,
      );

      expect(consoleSpy).not.toHaveBeenCalled();
    });
  });
});
