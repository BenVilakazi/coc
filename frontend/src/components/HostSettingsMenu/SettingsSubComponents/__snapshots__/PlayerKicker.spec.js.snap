// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerKicker rendering matches the snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
    type="button"
  >
    MAIN BUTTON
  </button>
</div>
`;
